### PR TITLE
AWS add vpc-endpoint filter for vpc

### LINKED ISSUE #5933: AWS - vpc-endpoint filter for vpc resource
**Is your feature request related to a problem? Please describe.**
Yes, we're trying to filter vpc based on their vpc endpoints. for example filter vpc's that have a s3 vpc endpoint

**Describe the solution you'd like**
I am thinking of creating a related resource filter that filters vpcs based on their vpc endpoints. 

for example 
```
policies:
  - name: s3-vpc-endpoint-enabled
    resource: vpc
    filters:
       - type: vpc-endpoint
          key: ServiceName
          value: com.amazonaws.us-east-1.s3
```

**Additional context**
The goal is to ensure certain vpc's contain certain vpc endpoints

### PR DESCRIPTION
**Pull request by anovis**
Adds vpc-endpoint filter for a vpc. 

for example
```
            policies:
              - name: s3-vpc-endpoint-enabled
                resource: vpc
                filters:
                  - type: vpc-endpoint
                    key: ServiceName
                    value: com.amazonaws.us-east-1.s3
```
closes #5933

### CODE REVIEW

File: c7n/resources/vpc.py [hunk 1]
```diff
@@ -2173,6 +2173,39 @@ class EndpointVpcFilter(net_filters.VpcFilter):
...
+    RelatedResource = "c7n.resources.vpc.VpcEndpoint"
+    RelatedIdsExpression = "VpcId"
+    AnnotationKey = "matched-vpc-endpoint"
+
+    schema = type_schema(
+        'vpc-endpoint',
+        rinherit=ValueFilter.schema)
+
+    def get_related_ids(self, resources):
+        vpc_set = set(jmespath.search(
```
Comment (JohnHillegass) [inline c7n/resources/vpc.py:2201]: nice work with this! 

There could be an opportunity for inheritance here by using `super().get_related_ids(resources)`.

File: c7n/resources/vpc.py [hunk 2]
```diff
@@ -2173,6 +2173,39 @@ class EndpointVpcFilter(net_filters.VpcFilter):
...
+            "[].%s" % self.RelatedIdsExpression, resources))
```
Comment (JohnHillegass) [inline c7n/resources/vpc.py:2202]: nice work with this! 

There could be an opportunity for inheritance here by calling `super()`.

Comment (anovis) [reply c7n/resources/vpc.py]: good call! will update

File: tests/test_vpc.py [hunk 3]
```diff
@@ -258,6 +258,25 @@ def test_dhcp_options_filter(self):
...
+                        "key": "ServiceName",
+                        "value": "com.amazonaws.us-east-1.s3",
+                    }
+                ],
+            },
+            session_factory=factory,
+        )
+        resources = p.run()
+        self.assertEqual(len(resources), 1)
+
```
Comment (JohnHillegass) [inline tests/test_vpc.py:279]: Would it make sense to add an assertion checking that the expected VPC endpoint was matched? 
e.g.
`"vpce-0bc8bcfcc74fa2db9" in resource[0]["c7n:matched-vpc-endpoint"]`

Comment (anovis) [reply tests/test_vpc.py:280]: yeah sounds good. will add.

**Comment by PratMis** [comment]
Looks like we can choose the subnets in a vpc to associate the vpc-endpoint to. For ex - If we have two subnets in a vpc, we can choose to grant one subnet access to the endpoint while blocking the other subnet. 
Feel like this filter should be at the subnet level than having it at the vpc level, since a vpc can have multiple subnets.

**Comment by JohnHillegass** [comment]
> Looks like we can choose the subnets in a vpc to associate the vpc-endpoint to. For ex - If we have two subnets in a vpc, we can choose to grant one subnet access to the endpoint while blocking the other.
> Feel like this filter should be at the subnet level than having it at the vpc level, since a vpc can have multiple subnets.

@PratMis Fair point! I had an impression the original intention was to simply check if a VPC endpoint connected to a VPC. I guess it depends on the implementation need. @anovis do you need to also look at the subnet level? I guess there could be a subtle distinction between `all vpcs must have a s3 vpc endpoint attached` and `all subnets must have a s3 vpc endpoint attached`

**Comment by anovis** [comment]
yeah makes sense. i can make the same filter for the subnet resource as well.

File: c7n/resources/vpc.py [hunk 4]
```diff
@@ -2173,6 +2173,38 @@ class EndpointVpcFilter(net_filters.VpcFilter):
...
+    schema = type_schema(
+        'vpc-endpoint',
+        rinherit=ValueFilter.schema)
+
+    def get_related_ids(self, resources):
+        vpc_set = super().get_related_ids(resources)
+        resource_manager = self.get_resource_manager()
+        model = resource_manager.get_model()
+
+        return {r[model.id] for r in resource_manager.resources() if r["VpcId"] in vpc_set}
```
Comment (kapilt) [inline c7n/resources/vpc.py:2205]: why is this refetching resources instead of using passed in resources?

File: c7n/resources/vpc.py [hunk 5]
```diff
@@ -2173,6 +2173,38 @@ class EndpointVpcFilter(net_filters.VpcFilter):
...
+    """
+    RelatedResource = "c7n.resources.vpc.VpcEndpoint"
+    RelatedIdsExpression = "VpcId"
+    AnnotationKey = "matched-vpc-endpoint"
+
+    schema = type_schema(
+        'vpc-endpoint',
+        rinherit=ValueFilter.schema)
+
+    def get_related_ids(self, resources):
```
Comment (kapilt) [inline c7n/resources/vpc.py:2200]: get related ids is generally intended to be a lightweight implementation, it gets called frequently, ie once per resource, doing an api call / or cache lookup isn't ideal for an implementation.

Comment (anovis) [reply c7n/resources/vpc.py]: yeah good call out. i had noticed that while writing the subnet tests. 

Instead I modified the get_related and process_resource, but not sure if that is the best way.

**Comment by kapilt** [comment]
so rather than duplicating the implementation for vpc and subnet, it would be better to have a related base class that addresses on the commonality. its distinct from the extant related base since its filtering the related resource to determine the related ids instead of using the policy resource to determine the ids. that feels like it has some reuse potential.

**Review by kapilt — APPROVED** [review]
lgtm, thanks

**Pull request merged**